File: 코딩테스트 고득점 Kit/DFSBFS/단어변환/단어 변환.py
import sys
sys.setrecursionlimit(51)


def solution(begin, target, words):
    answer = 51
    visited = [False] * len(words)

    def is_change(word1, word2):
        w = 0
        for i in range(len(word1)):
            if word1[i] != word2[i]:
                w += 1

        return True if w == 1 else False

    def DFS(cur, target, words, visited, depth):
        nonlocal answer

        # 도달 했다면
        if cur == target:
            answer = min(answer, depth)  # 최솟값 반환
            return

        for i in range(len(words)):
            # 변환 가능하고 미방문
            if is_change(cur, words[i]) and not visited[i]:
                visited[i] = True
                DFS(words[i], target, words, visited, depth + 1)
                visited[i] = False
        return

    if target in words:
        # dfs
        DFS(begin, target, words, visited, 0)
    else:
        answer = 0
    return answer

File: 코딩테스트 고득점 Kit/DFSBFS/단어변환/단어 변환.swift
import Foundation

func solution(_ begin:String, _ target:String, _ words:[String]) -> Int {
    var answer:Int = Int.max
    var visited:[Bool] = Array(repeating: false, count: words.count)
    
    func check(_ cur: [String], _ next: [String]) -> Bool {
        var cnt:Int = 0
        
        for i in 0..<cur.count {
            if cur[i] != next[i] {
                cnt += 1
            }
        }
        
        return cnt == 1 ? true : false
    }
    
    func DFS(_ cur: String, _ depth: Int) {
        if cur == target {
            answer = min(depth, answer)
            return
        }
        
        for i in 0..<words.count {
            if visited[i] {
                continue
            }
            
            if !check(cur.map{String($0)}, words[i].map{String($0)}) {
                continue
            }
            
            visited[i] = true
            DFS(words[i], depth+1)
            visited[i] = false
        }
        return
    }
    
    if words.contains(target) {
        DFS(begin, 0)
    } else {
        answer = 0
    }
    
    return answer
}